### PR TITLE
Rework onAppointmentFormOpening

### PR DESCRIPTION
**Pull request by pavel-2020**

### CODE REVIEW

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 1]
```diff
@@ -30,80 +30,114 @@ The form's instance; created only once - when the function is executed for the f
...
+The form items are organized into two groups:
+
+<table class="dx-table">
+    <tr>
+        <th>Group name</th>
+        <th>Description</th>
+    </tr>
+    <tr>
+        <td>mainGroup</td>
+        <td>Contains form fields that define appointment's main parameters (subject, start and end dates, etc.).</td>
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
        <td>Contains form fields that define main appointment parameters (subject, start and end dates, etc.).</td>
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 2]
```diff
@@ -30,80 +30,114 @@ The form's instance; created only once - when the function is executed for the f
...
+    </tr>
+    <tr>
+        <td>recurrenceGroup</td>
+        <td>Contains form fields that define appointment's recurrence parameters.</td>
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
        <td>Contains form fields that define appointment recurrence parameters.</td>
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 3]
```diff
@@ -30,80 +30,114 @@ The form's instance; created only once - when the function is executed for the f
...
-                });
-                if (!formItems.find(function(i) { return i.dataField === "location" })) {
+                text: "Start the project",
+                startDate: new Date(2020, 4, 24, 9, 10),
+                endDate: new Date(2020, 4, 24, 11, 20),
+            }],
+            currentDate: new Date(2020, 4, 24),
+
+            onAppointmentFormOpening: function(e) {
+                let form = e.form;
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
                const form = e.form;
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 4]
```diff
@@ -30,40 +30,64 @@ The form's instance; created only once - when the function is executed for the f
...
+        <td>recurrenceGroup</td>
+        <td>Contains form fields that define appointment recurrence parameters.</td>
+    </tr> 
+</table>
+
+You can add a custom item to any group or create an ungrouped item and display it under the groups, as shown in the following image:
+
+![DevExtreme Scheduler onAppointmentFormOpening](/images/UiWidgets/Scheduler_onAppointmentFormOpening.png)
+
+The code below adds a new form item (`phone`) to the `mainGroup` and creates an ungrouped item (`location`). Note that the array of [form items](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/#items) should be checked to ensure that it does not already contain an item with the same data field.
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
The code below adds a new form item (`phone`) to the `mainGroup` and creates an ungrouped item (`location`). Note that the array of [form items](/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/#items) should be checked to ensure that it does not already contain an item with the same data field. 
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 5]
```diff
@@ -74,33 +98,43 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+            text: "Start the project",
+            startDate: new Date(2020, 4, 24, 9, 10),
+            endDate: new Date(2020, 4, 24, 11, 20),
+        }];
+        currentDate = new Date(2020, 4, 24);
+        
+        onAppointmentFormOpening(e) {
+            const form = e.form;
+            let mainGroupItems = form.itemOption('mainGroup').items; 
+            if (!mainGroupItems.find(function(i) { return i.dataField === "author" })) {
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
            if (!mainGroupItems.find(function(i) { return i.dataField === "phone" })) {
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 6]
```diff
@@ -74,33 +98,43 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+                    colSpan: 2, 
+                    label: { text: "Phone Number" },
+                    editorType: "dxTextBox",
+                    dataField: "phone"
+                });
+                form.itemOption('mainGroup', 'items', mainGroupItems);
+            }
+    
+            let formItems = form.option("items"); 
+            if (!formItems.find(function(i) { return i.dataField === "checkList" })) {
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
            if (!formItems.find(function(i) { return i.dataField === "location" })) {
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 7]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
-        (onAppointmentFormOpening)="onAppointmentFormOpening($event)">
-    </dx-scheduler>
+##### Vue
+
+    <!-- tab: App.vue -->
+    <template>
+        <DxScheduler
+            :dataSource="schedulerData"
+            :currentDate="currentDate"
+            :on-appointment-form-opening="onAppointmentFormOpening">
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
            @appointment-form-opening="onAppointmentFormOpening"
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 8]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
-        [dataSource]="schedulerData"
-        [currentDate]="currentDate"
-        (onAppointmentFormOpening)="onAppointmentFormOpening($event)">
-    </dx-scheduler>
+##### Vue
+
+    <!-- tab: App.vue -->
+    <template>
+        <DxScheduler
+            :dataSource="schedulerData"
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
            :data-source="schedulerData"
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 9]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+            :currentDate="currentDate"
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
            :current-date="currentDate"
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 10]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+            :on-appointment-form-opening="onAppointmentFormOpening">
+        </DxScheduler>
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
        />
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 11]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+
+    <script>
+    import 'devextreme/dist/css/dx.common.css';
+    import 'devextreme/dist/css/dx.light.css';
+
+    import DxScheduler from 'devextreme-vue/scheduler';
+
+    export default {
+        components: {
+            DxScheduler,
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
            DxScheduler
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 12]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+                    endDate: new Date(2020, 4, 24, 11, 20),
+                }],
+                currentDate: new Date(2020, 4, 24)
+            }
+        },
+        methods: {
+            onAppointmentFormOpening(e) {
+                const form = e.form;
+                let mainGroupItems = form.itemOption('mainGroup').items; 
+                if (!mainGroupItems.find(function(i) { return i.dataField === "author" })) {
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
                if (!mainGroupItems.find(function(i) { return i.dataField === "phone" })) {
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 13]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+                        colSpan: 2, 
+                        label: { text: "Phone Number" },
+                        editorType: "dxTextBox",
+                        dataField: "phone"
+                    });
+                    form.itemOption('mainGroup', 'items', mainGroupItems);
+                }
+        
+                let formItems = form.option("items"); 
+                if (!formItems.find(function(i) { return i.dataField === "checkList" })) {
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
                if (!formItems.find(function(i) { return i.dataField === "location" })) {
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 14]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+
+    import Scheduler from 'devextreme-react/scheduler';
+
+    class App extends React.Component {
+        currentDate = new Date(2020, 4, 24);
+        
+        onAppointmentFormOpening(e) {
+            const form = e.form;
+            let mainGroupItems = form.itemOption('mainGroup').items;
+            if (!mainGroupItems.find(function(i) { return i.dataField === "author" })) {
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
            if (!mainGroupItems.find(function(i) { return i.dataField === "phone" })) {
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 15]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+                    colSpan: 2, 
+                    label: { text: "Phone Number" },
+                    editorType: "dxTextBox",
+                    dataField: "phone"
+                });
+                form.itemOption('mainGroup', 'items', mainGroupItems);
+            }
+    
+            let formItems = form.option("items"); 
+            if (!formItems.find(function(i) { return i.dataField === "checkList" })) {
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
            if (!formItems.find(function(i) { return i.dataField === "location" })) {
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 16]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+                    dataField: "location"
+                });
+                form.option("items", formItems);
+            }
+        }
+
+        render() {
+            return (
+                <Scheduler 
+                    // dataSource={this.dataSource}
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 17]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+                    dataSource={dataSource}
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md:280]: `dataSource` is missing

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 18]
```diff
@@ -117,18 +151,120 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+                    currentDate={this.currentDate}
```
Comment (RomanTsukanov) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
                    defaultCurrentDate={this.currentDate}
```

**Review by RomanTsukanov — CHANGES_REQUESTED** [review]

**Review by RomanTsukanov — APPROVED** [review]

**Review by andrewmakarov — APPROVED** [review]

**Review by RomanTsukanov — APPROVED** [review]

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 19]
```diff
@@ -30,40 +30,69 @@ The form's instance; created only once - when the function is executed for the f
 Model data. Available only if you use Knockout.
 
 ##### field(e.popup): dxPopup
-<!-- Description goes here -->
+The instance of the popup that contains the form.
 
 ---
-The appointment details form contains the [Form](/concepts/05%20Widgets/Form/00%20Overview.md '/Documentation/Guide/Widgets/Form/Overview/') widget whose instance is passed to this function in the **form** field. Use the [widget's API](/api-reference/10%20UI%20Widgets/dxForm '/Documentation/ApiReference/UI_Widgets/dxForm/') to customize the appointment details form.
+The appointment details form is displayed inside a popup. These elements are implemented by the [Form](/Documentation/Guide/Widgets/Form/Overview/) and [Popup](/Documentation/Guide/Widgets/Popup/Overview/) widgets. Their instances are passed to **onAppointmentFormOpening** in the **form** and **popup** fields. Use the [Form API](/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/) and [Popup API](/Documentation/ApiReference/UI_Widgets/dxPopup/Configuration/) to customize the corresponding widget.
```
Comment (albertov05) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
The appointment details form is displayed inside a popup. These elements are implemented by the [Form](/Documentation/Guide/Widgets/Form/Overview/) and [Popup](/Documentation/Guide/Widgets/Popup/Overview/) widgets. Their instances are passed to the **onAppointmentFormOpening** function in the **form** and **popup** fields. Use the [Form API](/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/) and [Popup API](/Documentation/ApiReference/UI_Widgets/dxPopup/Configuration/) to customize the corresponding widget.
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 20]
```diff
@@ -30,40 +30,69 @@ The form's instance; created only once - when the function is executed for the f
...
 
-The following code shows how to use the **onAppointmentFormOpening** function to customize an existing form item (`startDate`) and add a new form item (`location`). Note that in the latter case, the array of [form items](/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/#items) should be checked to ensure that it does not already contain an item with the same data field.
+The form items are organized into two groups:
```
Comment (albertov05) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
Form items are organized into two groups:
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 21]
```diff
@@ -30,40 +30,69 @@ The form's instance; created only once - when the function is executed for the f
...
+        <td>mainGroup</td>
+        <td>Contains form fields that define main appointment parameters (subject, start and end dates, etc.).</td>
+    </tr>
+    <tr>
+        <td>recurrenceGroup</td>
+        <td>Contains form fields that define appointment recurrence parameters.</td>
+    </tr> 
+</table>
+
+You can add a custom item to any group or create an ungrouped item and display it under the groups, as shown in the following image:
```
Comment (albertov05) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md:55]: What do you mean by display an ungrouped item under groups? ("groups" sounds like this ungrouped item is displayed under several different groups.) Please clarify.

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md [hunk 22]
```diff
@@ -117,19 +161,134 @@ The following code shows how to use the **onAppointmentFormOpening** function to
...
+                    onAppointmentFormOpening={this.onAppointmentFormOpening}
+                />
+            );
+        }
+    }
+    export default App;
+
+---
 
+The `mainGroup` consists of two columns. To make a custom item span them both, set its [colSpan](/Documentation/ApiReference/UI_Widgets/dxForm/Item_Types/SimpleItem/#colSpan) to 2 as done in the code above. Apply the same setting to an ungrouped item if it should span the `mainGroup` and `recurrenceGroup`.
```
Comment (albertov05) [inline api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]:
```suggestion
The `mainGroup` consists of two columns. To make a custom item span both columns, set its [colSpan](/Documentation/ApiReference/UI_Widgets/dxForm/Item_Types/SimpleItem/#colSpan) to '2' - as shown in the code above. Apply the same setting to an ungrouped item if it should span the `mainGroup` and `recurrenceGroup`. 
```

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md (hunk 21 above)
Comment (pavel-2020) [reply api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md:55]: This ungrouped item is displayed under two groups.
Scheduler_onAppointmentFormOpening

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md (hunk 22 above)
Comment (pavel-2020) [reply api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]: Applied all these changes except for '2' - left it without quotes.

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md (hunk 21 above)
Comment (albertov05) [reply api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md:55]: Thank you for the clarification.

File: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md (hunk 22 above)
Comment (albertov05) [reply api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md]: Okay, thank you.

**Pull request merged**